api/filters/paging.go

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2018 The Service Manager Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package filters
+
+import (
+	"context"
+	"fmt"
+	"github.com/Peripli/service-manager/pkg/query"
+	"github.com/Peripli/service-manager/pkg/util"
+	"github.com/Peripli/service-manager/pkg/web"
+	"net/http"
+	"strconv"
+)
+
+// PagingFilterName is the name of the paging filter
+const PagingFilterName = "PagingFilter"
+
+// pagingFilter is filter that adds paging criteria to request
+type pagingFilter struct {
+	DefaultPageSize int
+	MaxPageSize     int
+}
+
+// NewPagingFilter returns new paging filter for given default and max page sizes
+func NewPagingFilter(defaultPageSize, maxPageSize int) web.Filter {
+	return &pagingFilter{
+		DefaultPageSize: defaultPageSize,
+		MaxPageSize:     maxPageSize,
+	}
+}
+
+// Name returns the name of the paging filter
+func (*pagingFilter) Name() string {
+	return PagingFilterName
+}
+
+// Run represents the paging middleware function that processes the request and configures needed criteria.
+func (pf *pagingFilter) Run(req *web.Request, next web.Handler) (*web.Response, error) {
+	ctx := req.Context()
+	maxItems := req.URL.Query().Get("max_items")
+	limit, err := pf.validateMaxItemsQuery(maxItems)
+	if err != nil {
+		return nil, err
+	}
+	ctx = context.WithValue(ctx, "limit", limit)
+	if limit > 0 {
+		ctx, err = query.AddCriteria(ctx, query.LimitResultBy(limit))
```

Comment:
Currently there are two layers in the implementation that i think miss proper validation and can kind of mess up the limitting

1. If another filter adds a limit criteria after this one, this one will be silently overriden 
2. in the pgQuery builder (postgres implementation) there is a method of adding limit that currently override the limit everytime limit is set by it.
3. Order by created_at is currently hardcoded in an ugly place in the code, maybe now as it will be required for paging we can refactor it slightly